### PR TITLE
Add addPrior method to NonlinearFactorGraph and corresponding unit test.

### PR DESCRIPTION
**Pull request by Alescontrela**
This PR adds the `NonlinearFactorGraph::addPrior<T>` method, intending to reduce the verbosity of the code required to add a `PriorFactor`.

**Before:**
```c++
gtsam::Pose3 prior = ...;
graph.add(gtsam::PriorFactor<gtsam::Pose3>(key, prior, model));
```

**After:**
```c++
gtsam::Pose3 prior = ...;
graph.addPrior(key, prior, model);
```

If these changes look useful to the reviewers, I can move to replace the original instances of `graph.add(gtsam::PriorFactor<...>)(...))` with the new method.

---
This change is [Reviewable](https://reviewable.io/reviews/borglab/gtsam/280)

### CODE REVIEW

File: gtsam/nonlinear/NonlinearFactorGraph.h [hunk 1]
```diff
@@ -202,6 +203,29 @@ namespace gtsam {
...
+    /**
+     * Directly add a PriorFactor to the factor graph.
+     * @param key    Variable key
+     * @param prior  The variable's prior value
+     * @param model  Noise model for prior factor
+     */
+    template<typename T>
+    void addPrior(Key key, const T& prior,
+                  const SharedNoiseModel& model = nullptr) {
+      push_back(boost::make_shared<PriorFactor<T>>(key, prior, model));
```
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactorGraph.h:215]: emplace_back

File: gtsam/nonlinear/NonlinearFactorGraph.h [hunk 2]
```diff
@@ -202,6 +203,29 @@ namespace gtsam {
...
+
+    /**
+     * Directly add a PriorFactor to the factor graph.
+     * @param key         Variable key
+     * @param prior       The variable's prior value
+     * @param covariance  Covariance matrix.
+     */
+    template<typename T>
+    void addPrior(Key key, const T& prior, const Matrix& covariance) {
+      push_back(boost::make_shared<PriorFactor<T>>(key, prior, covariance));
```
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactorGraph.h:226]: emplace_back

**Comment by jlblancoc** [comment]
IMHO... this may break the generality of the FactorGraph class, requiring the inclusion of a header from the slam "module"... 
OTOH, it's really convenient to have. 
Perhaps this would be better by moving PriorFactor out of "slam"? (And adding a typedef there for backwards compatibility of course)

File: gtsam/nonlinear/NonlinearFactorGraph.h [hunk 3]
```diff
@@ -202,6 +203,29 @@ namespace gtsam {
...
+                  const SharedNoiseModel& model = nullptr) {
+      push_back(boost::make_shared<PriorFactor<T>>(key, prior, model));
+    }
+
+    /**
+     * Directly add a PriorFactor to the factor graph.
+     * @param key         Variable key
+     * @param prior       The variable's prior value
+     * @param covariance  Covariance matrix.
+     */
```
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactorGraph.h:223]: Document that smart noise model will pick right noise model (e.g. diagonal).

File: tests/testNonlinearFactorGraph.cpp [hunk 4]
```diff
@@ -242,6 +243,49 @@ TEST(testNonlinearFactorGraph, eliminate) {
...
+/* ************************************************************************* */
+TEST(testNonlinearFactorGraph, addPrior) {
+  Key k(0);
+
+  // Factor graph.
+  auto graph = NonlinearFactorGraph();
+
+  // Add a prior factor for key k.
+  auto model_double = noiseModel::Isotropic::Sigma(1, 1);
+  graph.addPrior<double>(k, 10.0, model_double);
```
Comment (dellaert) [inline tests/testNonlinearFactorGraph.cpp:255]: do we *need* the <double>. Should be inferred? Same for all tests below.

File: tests/testNonlinearFactorGraph.cpp [hunk 5]
```diff
@@ -242,6 +243,49 @@ TEST(testNonlinearFactorGraph, eliminate) {
...
+
+  // Clear the factor graph and values.
+  values.clear();
+  graph.erase(graph.begin(), graph.end());
+
+  // Add a Pose3 prior to the factor graph. Use a gaussian noise model by
+  // providing the covariance matrix.
+  Eigen::DiagonalMatrix<double, 6, 6> covariance_pose3;
+  covariance_pose3.setIdentity();
+  Pose3 pose{Rot3(), Point3()};
```
Comment (dellaert) [inline tests/testNonlinearFactorGraph.cpp:275]: Point3 will have undefined values

**Comment by dellaert** [comment]
> IMHO... this may break the generality of the FactorGraph class, requiring the inclusion of a header from the slam "module"...
> OTOH, it's really convenient to have.
> Perhaps this would be better by moving PriorFactor out of "slam"? (And adding a typedef there for backwards compatibility of course)

That's a good point! I support moving prior into nonlinear, with the typedef.

File: gtsam/nonlinear/NonlinearFactorGraph.h [hunk 6]
```diff
@@ -202,6 +203,29 @@ namespace gtsam {
       push_back(boost::make_shared<ExpressionFactor<T> >(R, z, h));
     }
 
+    /**
+     * Directly add a PriorFactor to the factor graph.
```
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearFactorGraph.h:207]: Add a note mentioning the convenience such as 

```
Convenience function to add PriorFactors specially.

File: gtsam/nonlinear/NonlinearFactorGraph.h (hunk 1 above)
Comment (Alescontrela) [reply gtsam/nonlinear/NonlinearFactorGraph.h]: Done 👍

File: gtsam/nonlinear/NonlinearFactorGraph.h (hunk 2 above)
Comment (Alescontrela) [reply gtsam/nonlinear/NonlinearFactorGraph.h]: Done 👍

File: tests/testNonlinearFactorGraph.cpp (hunk 4 above)
Comment (Alescontrela) [reply tests/testNonlinearFactorGraph.cpp]: Good point. I removed unnecessary decimals.

File: gtsam/nonlinear/NonlinearFactorGraph.h (hunk 3 above)
Comment (Alescontrela) [reply gtsam/nonlinear/NonlinearFactorGraph.h:227]: Done 👍

File: tests/testNonlinearFactorGraph.cpp (hunk 5 above)
Comment (Alescontrela) [reply tests/testNonlinearFactorGraph.cpp]: Gotcha. I added constructor arguments.

File: gtsam/nonlinear/NonlinearFactorGraph.h (hunk 6 above)
Comment (Alescontrela) [reply gtsam/nonlinear/NonlinearFactorGraph.h]: Done 👍

File: gtsam/nonlinear/NonlinearFactorGraph.h [hunk 7]
```diff
@@ -24,6 +24,7 @@
 #include <gtsam/geometry/Point2.h>
 #include <gtsam/nonlinear/NonlinearFactor.h>
 #include <gtsam/inference/FactorGraph.h>
+#include <gtsam/slam/PriorFactor.h>
```
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactorGraph.h:27]: So, this will have to change. Breaks Include DAG. See comments by Jose and me :-)

**Comment by dellaert** [comment]
> > IMHO... this may break the generality of the FactorGraph class, requiring the inclusion of a header from the slam "module"...
> > OTOH, it's really convenient to have.
> > Perhaps this would be better by moving PriorFactor out of "slam"? (And adding a typedef there for backwards compatibility of course)
> 
> That's a good point! I support moving prior into nonlinear, with the typedef.

PS probably no typedef needed, just an include in the old (slam) include file. The cop file would move, obviously...

**Comment by jlblancoc** [comment]
> > That's a good point! I support moving prior into nonlinear, with the typedef.
> 
> PS probably no typedef needed, just an include in the old (slam) include file. The cop file would move, obviously...

Ooops, sure! I was thinking of libraries where each subdirectory has its own namespace :-)
No typedef needed, for sure.

File: gtsam/nonlinear/NonlinearFactorGraph.h (hunk 7 above)
Comment (Alescontrela) [reply gtsam/nonlinear/NonlinearFactorGraph.h]: Yep! I changed all

`#include <gtsam/slam/PriorFactor.h>`

to

`#include <gtsam/nonlinear/PriorFactor.h>`

File: gtsam/slam/PriorFactor.h [hunk 8]
```diff
@@ -15,106 +15,6 @@
...
-	// Alignment, see https://eigen.tuxfamily.org/dox/group__TopicStructHavingEigenMembers.html
-	enum { NeedsToAlign = (sizeof(T) % 16) == 0 };
-  public:
-	EIGEN_MAKE_ALIGNED_OPERATOR_NEW_IF(NeedsToAlign)
-  };
-
-} /// namespace gtsam
+// Note: PriorFactor has been moved to gtsam/nonlinear. This file has been
+// left here for backwards compatibility.
+#include <gtsam/nonlinear/PriorFactor.h>
```
Comment (dellaert) [inline gtsam/slam/PriorFactor.h:20]: Missing new line

**Comment by Alescontrela** [comment]
@dellaert Yep I've been working on changing all old cases of `graph.emplace_shared<PriorFactor<...>>(...)` and `graph.add(PriorFactor<...>(...))` to `graph.addPrior<...>(...)`. It's a tedious process 😅 I should be done by tomorrow.

**Comment by dellaert** [comment]
Awesome :-) note to the extent possible you should omit the <T> part, it should be inferred from the argument type in most cases...

File: gtsam/slam/PriorFactor.h (hunk 8 above)
Comment (Alescontrela) [reply gtsam/slam/PriorFactor.h]: Added the newline 👍

File: cmake/example_cmake_find_gtsam/main.cpp [hunk 9]
```diff
@@ -69,7 +68,7 @@ int main(int argc, char** argv) {
   // 2a. Add a prior on the first pose, setting it to the origin
   // A prior factor consists of a mean and a noise model (covariance matrix)
   noiseModel::Diagonal::shared_ptr priorNoise = noiseModel::Diagonal::Sigmas(Vector3(0.3, 0.3, 0.1));
-  graph.emplace_shared<PriorFactor<Pose2> >(1, Pose2(0, 0, 0), priorNoise);
+  graph.addPrior<>(1, Pose2(0, 0, 0), priorNoise);
```
Comment (dellaert) [inline cmake/example_cmake_find_gtsam/main.cpp:71]: Why the <> ? Does it not compile without?

File: gtsam/geometry/triangulation.h [hunk 10]
```diff
@@ -22,7 +22,7 @@
 #include <gtsam/geometry/CameraSet.h>
 #include <gtsam/geometry/Pose2.h>
 #include <gtsam/slam/TriangulationFactor.h>
-#include <gtsam/slam/PriorFactor.h>
+#include <gtsam/nonlinear/PriorFactor.h>
```
Comment (dellaert) [inline gtsam/geometry/triangulation.h:25]: A forgotten case in this file?

File: gtsam/slam/tests/testOrientedPlane3Factor.cpp [hunk 11]
```diff
@@ -17,7 +17,7 @@
  */
 
 #include <gtsam/slam/OrientedPlane3Factor.h>
-#include <gtsam/slam/PriorFactor.h>
+#include <gtsam/nonlinear/PriorFactor.h>
```
Comment (dellaert) [inline gtsam/slam/tests/testOrientedPlane3Factor.cpp:20]: Forgotten case?

File: gtsam_unstable/dynamics/tests/testIMUSystem.cpp [hunk 12]
```diff
@@ -8,7 +8,7 @@
 #include <CppUnitLite/TestHarness.h>
 
 #include <gtsam/slam/BetweenFactor.h>
-#include <gtsam/slam/PriorFactor.h>
+#include <gtsam/nonlinear/PriorFactor.h>
```
Comment (dellaert) [inline gtsam_unstable/dynamics/tests/testIMUSystem.cpp:11]: Forgotten case?

File: gtsam_unstable/nonlinear/tests/testConcurrentBatchFilter.cpp [hunk 13]
```diff
@@ -17,7 +17,7 @@
  */
 
 #include <gtsam_unstable/nonlinear/ConcurrentBatchFilter.h>
-#include <gtsam/slam/PriorFactor.h>
+#include <gtsam/nonlinear/PriorFactor.h>
```
Comment (dellaert) [inline gtsam_unstable/nonlinear/tests/testConcurrentBatchFilter.cpp:20]: Remove?

File: gtsam_unstable/nonlinear/tests/testConcurrentIncrementalSmootherGN.cpp [hunk 14]
```diff
@@ -17,7 +17,7 @@
  */
 
 #include <gtsam_unstable/nonlinear/ConcurrentIncrementalSmoother.h>
-#include <gtsam/slam/PriorFactor.h>
+#include <gtsam/nonlinear/PriorFactor.h>
```
Comment (dellaert) [inline gtsam_unstable/nonlinear/tests/testConcurrentIncrementalSmootherGN.cpp:20]: Remove>

File: gtsam_unstable/slam/tests/testMultiProjectionFactor.cpp [hunk 15]
```diff
@@ -16,7 +16,7 @@
  *  @date Nov 2009
  */
 
-#include <gtsam/slam/PriorFactor.h>
+#include <gtsam/nonlinear/PriorFactor.h>
```
Comment (dellaert) [inline gtsam_unstable/slam/tests/testMultiProjectionFactor.cpp:19]: Forgotten case?

File: cmake/example_cmake_find_gtsam/main.cpp (hunk 9 above)
Comment (Alescontrela) [reply cmake/example_cmake_find_gtsam/main.cpp]: I mistakenly thought the <> were still necessary even for deduced template arguments. I've replaced all `addPrior<>` with `addPrior`.

File: gtsam_unstable/dynamics/tests/testIMUSystem.cpp (hunk 12 above)
Comment (Alescontrela) [reply gtsam_unstable/dynamics/tests/testIMUSystem.cpp:11]: This file initializes a PriorFactor on line 49.

File: gtsam_unstable/nonlinear/tests/testConcurrentBatchFilter.cpp (hunk 13 above)
Comment (Alescontrela) [reply gtsam_unstable/nonlinear/tests/testConcurrentBatchFilter.cpp:20]: This file initializes multiple PriorFactors and reuses them in a single test.

File: gtsam_unstable/nonlinear/tests/testConcurrentIncrementalSmootherGN.cpp (hunk 14 above)
Comment (Alescontrela) [reply gtsam_unstable/nonlinear/tests/testConcurrentIncrementalSmootherGN.cpp:20]: This file initializes multiple PriorFactors and reuses them in a single test.

File: gtsam_unstable/slam/tests/testMultiProjectionFactor.cpp (hunk 15 above)
Comment (Alescontrela) [reply gtsam_unstable/slam/tests/testMultiProjectionFactor.cpp:19]: Done 👍

File: gtsam/geometry/triangulation.h (hunk 10 above)
Comment (Alescontrela) [reply gtsam/geometry/triangulation.h:25]: Done 👍

File: gtsam/slam/tests/testOrientedPlane3Factor.cpp (hunk 11 above)
Comment (Alescontrela) [reply gtsam/slam/tests/testOrientedPlane3Factor.cpp:20]: Done 👍

**Comment by dellaert** [comment]
Travis errored die to a time-out in one of the test builds. Maybe make another small commit to re-trigger, could be that with all these changes the ccache has to be re-populated.

**Comment by Alescontrela** [comment]
@dellaert Do you think this PR is ready to merge?

**Comment by dellaert** [comment]
> @dellaert Do you think this PR is ready to merge?

I merged :-)

**Pull request merged**